File: OpenDentBusiness/TableTypes/Def.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Text;
using System.Xml.Serialization;

namespace OpenDentBusiness {
	///<summary>The info in the definition table is used by other tables extensively.  Almost every table in the database links to definition.  Almost all links to this table will be to a DefNum.  Using the DefNum, you can find any of the other fields of interest, usually the ItemName.  Make sure to look at the Defs class to see how the definitions are used.  Loaded into memory ahead of time for speed.  Some types of info such as operatories started out life in this table, but then got moved to their own table when more complexity was needed.</summary>
	[Serializable]
	[CrudTable(TableName="definition")]
	public class Def:TableBase {
		///<summary>Primary key.</summary>
		[CrudColumn(IsPriKey=true)]
		public long DefNum;
		///<summary>Enum:DefCat</summary>
		public DefCat Category;
		///<summary>Order that each item shows on various lists. 0-indexed.</summary>
		public int ItemOrder;
		///<summary>Each category is a little different.  This field is usually the common name of the item.</summary>
		public string ItemName;
		///<summary>This field can be used to store extra info about the item.</summary>
		public string ItemValue;
		///<summary>Some categories include a color option.</summary>
		[XmlIgnore]
		public Color ItemColor;
		///<summary>If hidden, the item will not show on any list, but can still be referenced.</summary>
		public bool IsHidden;

		///<summary>Used only for serialization purposes</summary>
		[XmlElement("ItemColor",typeof(int))]
		public int ItemColorXml {
			get {
				return ItemColor.ToArgb();
			}
			set {
				ItemColor = Color.FromArgb(value);
			}
		}

		///<summary>Returns a copy of the def.</summary>
		public Def Copy() {
			return (Def)MemberwiseClone();
		}

		/*
		public Def(){

		}

		public Def(long defNum,DefCat category,int itemOrder,string itemName,string itemValue,Color itemColor,bool isHidden){
			DefNum=defNum;
			Category=category;
			ItemOrder=itemOrder;
			ItemName=itemName;
			ItemValue=itemValue;
			ItemColor=itemColor;
			IsHidden=isHidden;
		}*/

	}

	///<summary>Definition Category. Go to the definition setup window in the program to see how each of these categories is used.</summary>
	public enum DefCat {
		///<summary>0- Colors to display in Account module.</summary>
		AccountColors,
		///<summary>1- Adjustment types.</summary>
		AdjTypes,
		///<summary>2- Appointment confirmed types.</summary>
		ApptConfirmed,
		///<summary>3- Procedure quick add list for appointments.</summary>
		ApptProcsQuickAdd,
		///<summary>4- Billing types.</summary>
		BillingTypes,
		///<summary>5- Not used.</summary>
		ClaimFormats,
		///<summary>6- Not used.</summary>
		DunningMessages,
		///<summary>7- Not used.</summary>
		FeeSchedNamesOld,
		///<summary>8- Medical notes for quick paste.</summary>
		MedicalNotes,
		///<summary>9- No longer used</summary>
		OperatoriesOld,
		///<summary>10- Payment types.</summary>
		PaymentTypes,
		///<summary>11- Procedure code categories.</summary>
		ProcCodeCats,
		///<summary>12- Progress note colors.</summary>
		ProgNoteColors,
		///<summary>13- Statuses for recall, unscheduled, and next appointments.</summary>
		RecallUnschedStatus,
		///<summary>14- Service notes for quick paste.</summary>
		ServiceNotes,
		///<summary>15- Discount types.</summary>
		DiscountTypes,
		///<summary>16- Diagnosis types.</summary>
		Diagnosis,
		///<summary>17- Colors to display in the Appointments module.</summary>
		AppointmentColors,
		///<summary>18- Image categories.</summary>
		ImageCats,
		///<summary>19- Quick add notes for the ApptPhoneNotes, which is getting phased out.</summary>
		ApptPhoneNotes,
		///<summary>20- Treatment plan priority names.</summary>
		TxPriorities,
		///<summary>21- Miscellaneous color options.</summary>
		MiscColors,
		///<summary>22- Colors for the graphical tooth chart.</summary>
		ChartGraphicColors,
		///<summary>23- Categories for the Contact list.</summary>
		ContactCategories,
		///<summary>24- Categories for Letter Merge.</summary>
		LetterMergeCats,
		///<summary>25- Types of Schedule Blockouts.</summary>
		BlockoutTypes,
		///<summary>26- Categories of procedure buttons in Chart module</summary>
		ProcButtonCats,
		///<Summary>27- Types of commlog entries.</Summary>
		CommLogTypes,
		///<summary>28- Categories of Supplies</summary>
		SupplyCats,
		///<summary>29- Types of unearned income used in accrual accounting.</summary>
		PaySplitUnearnedType,
		///<summary>30- Prognosis types.</summary>
		Prognosis
	}

	



}

File: OpenDentalWeb/OpenDental/src/com/opendental/odweb/client/tabletypes/Def.java
package com.opendental.odweb.client.tabletypes;

import com.google.gwt.xml.client.Document;
import com.opendental.odweb.client.remoting.Serializing;

//DO NOT MAKE CHANGES TO THIS FILE.  THEY WILL GET OVERWRITTEN BY THE CRUD.
public class Def {
		/** Primary key. */
		public int DefNum;
		/** Enum:DefCat */
		public DefCat Category;
		/** Order that each item shows on various lists. 0-indexed. */
		public int ItemOrder;
		/** Each category is a little different.  This field is usually the common name of the item. */
		public String ItemName;
		/** This field can be used to store extra info about the item. */
		public String ItemValue;
		/** Some categories include a color option. */
		public int ItemColor;
		/** If hidden, the item will not show on any list, but can still be referenced. */
		public boolean IsHidden;

		/** Deep copy of object. */
		public Def deepCopy() {
			Def def=new Def();
			def.DefNum=this.DefNum;
			def.Category=this.Category;
			def.ItemOrder=this.ItemOrder;
			def.ItemName=this.ItemName;
			def.ItemValue=this.ItemValue;
			def.ItemColor=this.ItemColor;
			def.IsHidden=this.IsHidden;
			return def;
		}

		/** Serialize the object into XML. */
		public String serialize() {
			StringBuilder sb=new StringBuilder();
			sb.append("<Def>");
			sb.append("<DefNum>").append(DefNum).append("</DefNum>");
			sb.append("<Category>").append(Category.ordinal()).append("</Category>");
			sb.append("<ItemOrder>").append(ItemOrder).append("</ItemOrder>");
			sb.append("<ItemName>").append(Serializing.escapeForXml(ItemName)).append("</ItemName>");
			sb.append("<ItemValue>").append(Serializing.escapeForXml(ItemValue)).append("</ItemValue>");
			sb.append("<ItemColor>").append(ItemColor).append("</ItemColor>");
			sb.append("<IsHidden>").append((IsHidden)?1:0).append("</IsHidden>");
			sb.append("</Def>");
			return sb.toString();
		}

		/** Sets all the variables on this object based on the values in the XML document.  Variables that are not in the XML document will be null or their default values.
		 * @param doc A parsed XML document.  Must be valid XML.  Does not need to contain a node for every variable on this object.
		 * @throws Exception DeserializeFromXml is entirely encased in a try catch and will throw exceptions if anything goes wrong. */
		public void deserialize(Document doc) throws Exception {
			try {
				if(Serializing.getXmlNodeValue(doc,"DefNum")!=null) {
					DefNum=Integer.valueOf(Serializing.getXmlNodeValue(doc,"DefNum"));
				}
				if(Serializing.getXmlNodeValue(doc,"Category")!=null) {
					Category=DefCat.values()[Integer.valueOf(Serializing.getXmlNodeValue(doc,"Category"))];
				}
				if(Serializing.getXmlNodeValue(doc,"ItemOrder")!=null) {
					ItemOrder=Integer.valueOf(Serializing.getXmlNodeValue(doc,"ItemOrder"));
				}
				if(Serializing.getXmlNodeValue(doc,"ItemName")!=null) {
					ItemName=Serializing.getXmlNodeValue(doc,"ItemName");
				}
				if(Serializing.getXmlNodeValue(doc,"ItemValue")!=null) {
					ItemValue=Serializing.getXmlNodeValue(doc,"ItemValue");
				}
				if(Serializing.getXmlNodeValue(doc,"ItemColor")!=null) {
					ItemColor=Integer.valueOf(Serializing.getXmlNodeValue(doc,"ItemColor"));
				}
				if(Serializing.getXmlNodeValue(doc,"IsHidden")!=null) {
					IsHidden=(Serializing.getXmlNodeValue(doc,"IsHidden")=="0")?false:true;
				}
			}
			catch(Exception e) {
				throw new Exception("Error deserializing Def: "+e.getMessage());
			}
		}

		/** Definition Category. Go to the definition setup window in the program to see how each of these categories is used. */
		public enum DefCat {
			/** 0- Colors to display in Account module. */
			AccountColors,
			/** 1- Adjustment types. */
			AdjTypes,
			/** 2- Appointment confirmed types. */
			ApptConfirmed,
			/** 3- Procedure quick add list for appointments. */
			ApptProcsQuickAdd,
			/** 4- Billing types. */
			BillingTypes,
			/** 5- Not used. */
			ClaimFormats,
			/** 6- Not used. */
			DunningMessages,
			/** 7- Not used. */
			FeeSchedNamesOld,
			/** 8- Medical notes for quick paste. */
			MedicalNotes,
			/** 9- No longer used */
			OperatoriesOld,
			/** 10- Payment types. */
			PaymentTypes,
			/** 11- Procedure code categories. */
			ProcCodeCats,
			/** 12- Progress note colors. */
			ProgNoteColors,
			/** 13- Statuses for recall, unscheduled, and next appointments. */
			RecallUnschedStatus,
			/** 14- Service notes for quick paste. */
			ServiceNotes,
			/** 15- Discount types. */
			DiscountTypes,
			/** 16- Diagnosis types. */
			Diagnosis,
			/** 17- Colors to display in the Appointments module. */
			AppointmentColors,
			/** 18- Image categories. */
			ImageCats,
			/** 19- Quick add notes for the ApptPhoneNotes, which is getting phased out. */
			ApptPhoneNotes,
			/** 20- Treatment plan priority names. */
			TxPriorities,
			/** 21- Miscellaneous color options. */
			MiscColors,
			/** 22- Colors for the graphical tooth chart. */
			ChartGraphicColors,
			/** 23- Categories for the Contact list. */
			ContactCategories,
			/** 24- Categories for Letter Merge. */
			LetterMergeCats,
			/** 25- Types of Schedule Blockouts. */
			BlockoutTypes,
			/** 26- Categories of procedure buttons in Chart module */
			ProcButtonCats,
			/**  */
			CommLogTypes,
			/** 28- Categories of Supplies */
			SupplyCats,
			/** 29- Types of unearned income used in accrual accounting. */
			PaySplitUnearnedType,
			/** 30- Prognosis types. */
			Prognosis,
			/** 31- Custom Tracking, statuses such as 'review', 'hold', 'riskmanage', etc. */
			ClaimCustomTracking
		}


}
